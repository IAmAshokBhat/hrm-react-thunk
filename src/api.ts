import {
  ILoginData,
  INewAttendance,
  INewLeave,
  IRegisterAPI
} from './constants';
import { IDetailsAPIReq, IDetailsAPIReqWithDates } from './contracts';
import { fetchWithToken, getToken } from './utils';

const headers = {
  'X-JWT-Token': getToken()
};
const API_URL = process.env.REACT_APP_API_URL;

export const getUserListAPI = () => fetchWithToken('user/list');

export const getUserDetailsAPI = ({ id }: IDetailsAPIReq) =>
  fetchWithToken(`${API_URL}api/users/${id}`);

export const getAttendanceDetailsAPI = ({
  id,
  from,
  to
}: IDetailsAPIReqWithDates) =>
  fetchWithToken(`${API_URL}api/attendance/${id}?from=${from}&to=${to}`);

export const getLeaveBalanceAPI = ({ id }: IDetailsAPIReq) =>
  fetchWithToken(`${API_URL}api/leaves/${id}`);

export const getEventDetailsAPI = ({ id, from, to }: IDetailsAPIReqWithDates) =>
  fetchWithToken(`${API_URL}api/events/${id}?from=${from}&to=${to}`);

export const getAllLeavesAPI = ({ id }: IDetailsAPIReq) =>
  fetchWithToken(`${API_URL}api/all-leaves/${id}`);

export const applyLeaveAPI = (newLeave: INewLeave) =>
  fetchWithToken(`${API_URL}api/apply-leave`, {
    method: 'POST',
    body: JSON.stringify(newLeave)
  });

export const addAttendanceAPI = (newAttendance: INewAttendance) =>
  fetchWithToken(`${API_URL}api/attendance`, {
    method: 'POST',
    body: JSON.stringify(newAttendance)
  });

export const loginAPI = (userInfo: ILoginData) =>
  fetchWithToken(`${API_URL}login`, {
    method: 'POST',
    body: JSON.stringify(userInfo)
  });

export const getDesignationListAPI = () =>
  fetchWithToken(`${API_URL}designation`, { headers: headers });

export const registerAPI = (newUser: IRegisterAPI) =>
  fetchWithToken(`${API_URL}user`, {
    method: 'POST',
    body: JSON.stringify(newUser)
  });
