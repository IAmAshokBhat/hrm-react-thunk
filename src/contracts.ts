import { AnyAction, Dispatch } from 'redux';
import { TAny, TAnyObject } from './constants';

export type TFPromise = (params: TAny) => Promise<TAnyObject>;

export type TFPromiseAPI = (params?: TAny | TAnyObject) => Promise<TAnyObject>;

export type TFAction = (dispatch: Dispatch<AnyAction>) => Promise<void>;

/*
()
(1,2,3)
({key:value})
**
*/
