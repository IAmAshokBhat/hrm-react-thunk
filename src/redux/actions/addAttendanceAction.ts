import { Dispatch } from 'redux';
import { addAttendanceAPI } from '../../api';
import {
  ACTIONS,
  IAttendanceDetailsAPI,
  INewAttendance
} from '../../constants';

const setAddAttendanceFailureAction = () => ({
  type: `${ACTIONS.ADD_ATTENDANCE}_FAILURE`
});

const setAddAttendanceLoadingAction = () => ({
  type: `${ACTIONS.ADD_ATTENDANCE}_LOADING`
});

const addAttendanceSuccessAction = (payload: IAttendanceDetailsAPI) => ({
  type: `${ACTIONS.ADD_ATTENDANCE}_SUCCESS`,
  payload
});

export const addAttendanceAction =
  (newAttendance: INewAttendance) => (dispatch: Dispatch) => {
    dispatch(setAddAttendanceLoadingAction());
    return addAttendanceAPI(newAttendance)
      .then(
        (response) => response.json(),
        (error) => console.log('An error occurred.', error)
      )
      .then((json) => {
        const { ok } = json;
        if (!ok) {
          dispatch(setAddAttendanceFailureAction());
        }
        dispatch(addAttendanceSuccessAction(json));
      });
  };
