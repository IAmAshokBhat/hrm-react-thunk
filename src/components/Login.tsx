import {
  Button,
  FormControl,
  FormLabel,
  TextField,
  Typography
} from '@mui/material';
import Card from '@mui/material/Card';
import Grid from '@mui/material/Grid';
import { useSnackbar } from 'notistack';
import { useEffect } from 'react';
import { Field, Form } from 'react-final-form';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation, useNavigate } from 'react-router-dom';
import { ILoginData, NotificationType } from '../constants';
import { loginAction } from '../redux/actions';
import { selectLoginDetails } from '../redux/selectors';
import { isValidToken } from '../utils';

export const LoginPage = () => {
  const dispatch = useDispatch();
  const loginDetails = useSelector(selectLoginDetails);
  const navigate = useNavigate();
  const { enqueueSnackbar } = useSnackbar();
  const { state } = useLocation();

  const onSubmit = (values: ILoginData) => {
    dispatch(loginAction(values));
  };

  useEffect(() => {
    if (loginDetails && loginDetails.message) {
      enqueueSnackbar(loginDetails.message, {
        variant: NotificationType.ERROR
      });
    }

    if (isValidToken()) {
      // @ts-ignore
      navigate(state?.param || '/dashboard');
    }
  }, [loginDetails, enqueueSnackbar, state, navigate]);

  return (
    <Card style={{ maxWidth: 500, margin: '0 auto', padding: 20 }}>
      <Form
        onSubmit={onSubmit}
        render={({ handleSubmit }) => (
          <form onSubmit={handleSubmit}>
            <Grid container rowSpacing={2}>
              <Grid item xs={12}>
                <Typography textAlign="center" variant="h4">
                  HR Management System
                </Typography>
              </Grid>
              <Grid item xs={12}>
                <FormControl sx={{ width: '100%' }}>
                  <FormLabel>User ID</FormLabel>
                  <Field name="userId" width="100%">
                    {(props) => (
                      <TextField
                        name={props.input.name}
                        value={props.input.value}
                        onChange={props.input.onChange}
                        placeholder="User Id"
                      />
                    )}
                  </Field>
                </FormControl>
              </Grid>
              <Grid item xs={12}>
                <FormControl sx={{ width: '100%' }}>
                  <FormLabel>Password</FormLabel>
                  <Field name="password" width="100%">
                    {(props) => (
                      <TextField
                        name={props.input.name}
                        value={props.input.value}
                        onChange={props.input.onChange}
                        type="password"
                        placeholder="Password"
                      />
                    )}
                  </Field>
                </FormControl>
              </Grid>
              <Grid item xs={12} textAlign="right">
                <Button variant="contained" type="submit" fullWidth>
                  Login
                </Button>
              </Grid>
            </Grid>
          </form>
        )}
      />
    </Card>
  );
};
