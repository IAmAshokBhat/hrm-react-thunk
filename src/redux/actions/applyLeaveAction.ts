import { Dispatch } from 'redux';
import { applyLeaveAPI } from '../../api';
import { ACTIONS, IAttendanceDetailsAPI, INewLeave } from '../../constants';

const setApplyLeaveFailureAction = () => ({
  type: `${ACTIONS.APPLY_LEAVE}_FAILURE`
});

const setApplyLeaveLoadingAction = () => ({
  type: `${ACTIONS.APPLY_LEAVE}_LOADING`
});

const applyLeaveSuccessAction = (payload: IAttendanceDetailsAPI) => ({
  type: `${ACTIONS.APPLY_LEAVE}_SUCCESS`,
  payload
});

export const applyLeaveAction =
  (newLeave: INewLeave) => (dispatch: Dispatch) => {
    dispatch(setApplyLeaveLoadingAction());
    return applyLeaveAPI(newLeave)
      .then(
        (response) => response.json(),
        (error) => console.log('An error occurred.', error)
      )
      .then((json) => {
        const { ok } = json;
        if (!ok) {
          dispatch(setApplyLeaveFailureAction());
        }
        dispatch(applyLeaveSuccessAction(json));
      });
  };
