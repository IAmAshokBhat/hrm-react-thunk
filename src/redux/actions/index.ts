import { fetchUserDetailsAction } from './userInfoAction';
import { fetchUsersListAction } from './usersActions';
import {
  fetchAttendanceDetailsAction,
  fetchAllLeavesAction
} from './attedanceDetailsAction';
import { fetchLeaveBalanceAction } from './leaveBalanceAction';
import { fetchEventDetailsAction } from './eventDetailsAction';
import { applyLeaveAction } from './applyLeaveAction';
import { addAttendanceAction } from './addAttendanceAction';
import { loginAction } from './loginAction';
import { registerAction } from './registerAction';
import { fetchDesignationListAction } from './designationListAction';

export {
  fetchUsersListAction as fetchUsersAction,
  fetchUserDetailsAction,
  fetchAttendanceDetailsAction,
  fetchLeaveBalanceAction,
  fetchEventDetailsAction,
  applyLeaveAction,
  addAttendanceAction,
  loginAction,
  registerAction,
  fetchDesignationListAction,
  fetchAllLeavesAction
};
