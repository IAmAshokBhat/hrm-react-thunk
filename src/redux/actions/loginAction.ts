import { Dispatch } from 'redux';
import { loginAPI } from '../../api';
import { ACTIONS, ILoginAPI, ILoginData } from '../../constants';

const setLoginFailureAction = () => ({
  type: ACTIONS.LOGIN_FAILURE
});

const setLoginLoadingAction = () => ({
  type: ACTIONS.LOGIN_LOADING
});

const loginSuccessAction = (payload: ILoginAPI) => ({
  type: ACTIONS.LOGIN_SUCCESS,
  payload
});

export const loginAction = (userInfo: ILoginData) => (dispatch: Dispatch) => {
  dispatch(setLoginLoadingAction());
  return loginAPI(userInfo)
    .then((response) => {
      if (response.status !== 200) {
        dispatch(setLoginFailureAction());
      } else {
        return response.json();
      }
    })
    .then((json: ILoginAPI) => {
      if (json) {
        const { token, expires } = json;

        localStorage.setItem('userToken', token);
        localStorage.setItem('expires', expires + '');
        localStorage.setItem('userId', userInfo.userId + '');
        dispatch(loginSuccessAction(json));
      }
    });
};
