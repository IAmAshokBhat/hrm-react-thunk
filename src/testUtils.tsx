import { render, RenderOptions, waitFor } from '@testing-library/react';
import { SnackbarProvider } from 'notistack';
import { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { applyMiddleware, createStore, Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IDefaultRoot, TAny, TAnyObject } from './constants';
import { TFAction } from './contracts';
import { reducers } from './redux/reducers';
import {
  defaultArrayState,
  defaultNullState,
  defaultState
} from './redux/reducers/__tests__/fixtures';

export const middlewares = [thunk];

export const createStoreWithMiddlewares = (
  initialState: TAnyObject = {}
): Store => {
  return createStore(reducers, initialState, applyMiddleware(...middlewares));
};
interface IRenderWithProviderOptions {
  initialStoreState?: TAnyObject;
  renderFunction?: TAny;
  options?: RenderOptions;
  route?: string;
  path?: string;
}

export const renderWithProvider = (
  ui: ReactElement,
  renderOptions?: IRenderWithProviderOptions
) => {
  const {
    initialStoreState = {},
    renderFunction = render,
    options = {},
    route = '',
    path = ''
  } = renderOptions || {};

  const appStore = createStoreWithMiddlewares(initialStoreState);

  let wrappedUi = (
    <SnackbarProvider
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'right'
      }}
    >
      <Provider store={appStore}>{ui}</Provider>
    </SnackbarProvider>
  );

  if (route) {
    window.history.pushState({}, '', route);

    wrappedUi = (
      <Provider store={appStore}>
        <BrowserRouter>
          <Routes>
            <Route path={path} element={ui} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  }

  return renderFunction(wrappedUi, options);
};

type TFApiCall = (...params: TAny) => Promise<Response>;

export interface IResponseMock {
  headers?: object;
  ok?: boolean;
  status?: number;
  statusText?: string;
  type?: ResponseType;
  url?: string;
  body?: TAny | null;
  bodyUsed?: boolean;
  blob?: () => Promise<TAny>;
}

const dummyPromise = () => new Promise((resolve) => resolve('dummy'));

export class ResponseMock {
  headers: object;

  ok: boolean;

  status: number;

  statusText: string;

  type: ResponseType;

  url: string;

  body: ReadableStreamDefaultReadResult<Uint8Array> | null;

  bodyUsed: boolean;

  blob: () => Promise<TAny>;

  constructor(props: IResponseMock = {}) {
    this.headers = props.headers || {};
    this.status = props.status || 200;
    this.ok = props.ok || this.status === 200;
    this.statusText = props.statusText || 'OK';
    this.type = props.type || 'basic';
    this.url = props.url || 'http://localhost';
    this.body = props.body;
    this.bodyUsed = props.bodyUsed || this.body !== null;
    this.blob = props.blob || dummyPromise;
  }

  json: () => Promise<TAny> = () => {
    const body = this.body || '""';

    try {
      return Promise.resolve(JSON.parse(body as string));
    } catch {
      return Promise.resolve(body);
    }
  };

  text: () => Promise<string> = () => Promise.resolve(this.statusText);
}

export const mockAPICall = (func: TFApiCall, response: IResponseMock) =>
  (func as jest.Mock).mockResolvedValue(new ResponseMock(response));

export const generateActionTests = async (
  actionName: string,
  action: TFAction,
  api: TAny,
  actionPayload: TAny = []
) => {
  const response = new ResponseMock({ ok: true, body: { data: [] } });

  const mockStore = configureStore(middlewares);

  (api as jest.Mock).mockReturnValue(Promise.resolve(response));
  const store = mockStore({});

  //TODO: Fix TS error
  // @ts-ignore
  await store.dispatch(action);

  await waitFor(() => {
    expect(store.getActions()).toEqual([
      {
        type: `${actionName}_LOADING`
      },
      {
        type: `${actionName}_SUCCESS`,
        //TODO: Accept action params from function params
        payload: actionPayload
      }
    ]);
  });
};

export const generateReducerTests = (
  reducerName: string,
  reducer: TAny,
  selector: TAny,
  actionName: string,
  successValue?: TAny,
  defaultValue?: TAny,
  expectedValue?: TAny,
  reducerLabel?: string
) => {
  describe(`for ${reducerName}`, () => {
    const reducerLbl = reducerLabel
      ? reducerLabel
      : reducerName.replace('Reducer', '');
    it(`${reducerName} reducer loading`, () => {
      reducer(undefined, {
        type: `${actionName}_LOADING`,
        payload: null
      });
      // TODO: Check why its failing
      // expect(selector(defaultState).loading).toEqual(true);
      expect(selector(defaultState).value).toEqual(
        defaultValue === null ? defaultValue : []
      );
      expect(selector(defaultState).error).toEqual(null);
    });

    it(`${reducerName} reducer success`, () => {
      const successState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: expectedValue ? expectedValue : successValue
        }
      };
      reducer(defaultArrayState, {
        type: `${actionName}_SUCCESS`,
        payload: successValue
      });

      expect(selector(successState).error).toEqual(null);
      expect(selector(successState).loading).toEqual(false);
      expect(selector(successState).value).toEqual(
        expectedValue ? expectedValue : successValue
      );
    });

    it(`${reducerName} reducer error`, () => {
      const error = { message: '', status: 1, ok: false };
      const failedState: IDefaultRoot = {
        ...defaultState,
        [reducerLbl]: {
          ...defaultNullState,
          value: [],
          error
        }
      };

      reducer(undefined, {
        type: `${actionName}_FAILURE`,
        payload: error
      });

      expect(selector(failedState).loading).toEqual(false);
      expect(selector(failedState).value).toEqual([]);
      expect(selector(failedState).error).toEqual(error);
    });
  });
};
