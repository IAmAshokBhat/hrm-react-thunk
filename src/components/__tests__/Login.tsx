import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SnackbarProvider } from 'notistack';
import { loginAPI } from '../../api';
import { TAnyObject } from '../../contracts';
import * as Actions from '../../redux/actions/loginAction';
import { mockAPICall, renderWithProvider } from '../../testUtils';
import { LoginPage } from '../Login';
import { defaultState } from './fixtures';

const mockEnqueue = jest.fn();

jest.mock('../../api', () => ({
  loginAPI: jest.fn()
}));

jest.mock('notistack', () => ({
  ...jest.requireActual('notistack'),
  useSnackbar: () => {
    return {
      enqueueSnackbar: mockEnqueue
    };
  }
}));

const loginResponse = {
  ok: true,
  body: {
    token: 'token',
    issued: 1644407868611,
    expires: 1644408768611,
    userId: '5',
    loginStatus: 1,
    message: 'success',
    status: 200
  }
};

describe('Login Page', () => {
  const generateNode = (
    initialStoreState: TAnyObject = defaultState,
    route = '/login'
  ) =>
    renderWithProvider(
      <SnackbarProvider>
        <LoginPage />
      </SnackbarProvider>,
      {
        route,
        path: '/login',
        initialStoreState
      }
    );

  it('should render and match snapshot', () => {
    const { container } = generateNode();

    expect(container).toMatchSnapshot();
  });

  it('should login when valid username and password is entered', async () => {
    const loginActionSpy = jest.spyOn(Actions, 'loginAction');

    mockAPICall(loginAPI, loginResponse);

    generateNode();

    userEvent.type(screen.getByPlaceholderText(/User Id/i), '5');
    userEvent.type(screen.getByPlaceholderText(/Password/i), 'I@mCrazy');
    await waitFor(() =>
      userEvent.click(screen.getByRole('button', { name: 'Login' }))
    );

    await waitFor(() =>
      expect(loginActionSpy).toHaveBeenCalledWith({
        password: 'I@mCrazy',
        userId: '5'
      })
    );

    await waitFor(() =>
      expect(loginAPI).toHaveBeenCalledWith({
        password: 'I@mCrazy',
        userId: '5'
      })
    );
  });
});
