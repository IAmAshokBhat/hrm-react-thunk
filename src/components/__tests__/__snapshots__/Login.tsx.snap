// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should render and match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    style="max-width: 500px; margin: 0px auto; padding: 20px;"
  >
    <form>
      <div
        class="MuiGrid-root MuiGrid-container css-1jzi3sl-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-149xwtl-MuiTypography-root"
          >
            HR Management System
          </h4>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiFormLabel-colorPrimary css-u4tvz2-MuiFormLabel-root"
            >
              User ID
            </label>
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-1"
                  name="userId"
                  placeholder="User Id"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-nnbavb"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiFormLabel-colorPrimary css-u4tvz2-MuiFormLabel-root"
            >
              Password
            </label>
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-2"
                  name="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-nnbavb"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13ezsq2-MuiGrid-root"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
